server.go

```diff
@@ -296,19 +305,53 @@ type Server struct {
 	DecorateReader DecorateReader
 	// DecorateWriter is optional, allows customization of the process that writes raw DNS messages.
 	DecorateWriter DecorateWriter
+	// Graceful shutdown handling
+	running int32
+	// Messages queue
+	queue chan *message
+	// Number of workers, if set to zero - use spawn goroutines instead
+	Workers int
+	// Size of DNS requests queue
+	QueueSize int
+}
+
+func (srv *Server) start() {
+	atomic.StoreInt32(&(srv.running), int32(1))
+}
+
+func (srv *Server) stop() {
+	atomic.StoreInt32(&(srv.running), int32(0))
+}
 
-	// Shutdown handling
-	lock    sync.RWMutex
-	started bool
+func (srv *Server) isRunning() bool {
+	return atomic.LoadInt32(&(srv.running)) == 1
+}
+
+func (srv *Server) startWorkers() {
+	srv.queue = make(chan *message, srv.QueueSize)
+	for i := 0; i < srv.Workers; i++ {
+		go func() {
+			for msg := range srv.queue {
+				srv.serve(msg)
+			}
+		}()
+	}
 }
 
 // ListenAndServe starts a nameserver on the configured address in *Server.
 func (srv *Server) ListenAndServe() error {
-	srv.lock.Lock()
-	defer srv.lock.Unlock()
-	if srv.started {
+	if srv.isRunning() {
```

Comment:
Is this need in this PR? (Not opposing the change - but it seems to do the same thing as earlier code, or is there a bug fixed?)

[it does clear out of awful locking we had in server]

Can you make this a seperate PR?